Mathlib/Algebra/Order/Archimedean/Class.lean

```diff
@@ -0,0 +1,393 @@
+    · rw [← mk_left_eq_mk_mul hlt]
+      congr
+      apply le_antisymm (Finset.le_min' _ _ _ ?_) (Finset.min'_le _ _ (by simp))
+      intro y hy
+      obtain rfl|hmem := Finset.mem_cons.mp hy
+      · rfl
+      · refine (lt_of_lt_of_le ?_ (Finset.min'_le _ _ hmem)).le
+        apply (hmono.lt_iff_lt (by simp) hminmem).mp
+        rw [ih] at hlt
+        exact hlt
+    · rw [mul_comm, ← mk_left_eq_mk_mul hlt, ih]
+      congr 2
+      refine le_antisymm (Finset.le_min' _ _ _ ?_) (Finset.min'_le _ _ hminmem)
+      intro y hy
+      obtain rfl|hmem := Finset.mem_cons.mp hy
+      · apply ((hmono.lt_iff_lt hminmem (by simp)).mp ?_).le
+        rw [ih] at hlt
+        exact hlt
+      · exact Finset.min'_le _ _ hmem
+
+@[to_additive]
+theorem lt_of_mk_lt_mk_of_one_le (h : mk a < mk b) (hpos : 1 ≤ a) : b < a := by
+  obtain h := (mk_lt_mk).mp h 1
+  rw [pow_one, mabs_lt, mabs_eq_self.mpr hpos] at h
+  exact h.2
+
+@[to_additive]
+theorem lt_of_mk_lt_mk_of_le_one (h : mk a < mk b) (hneg : a ≤ 1) : a < b := by
+  obtain h := (mk_lt_mk).mp h 1
+  rw [pow_one, mabs_lt, mabs_eq_inv_self.mpr hneg, inv_inv] at h
+  exact h.1
+
+@[to_additive]
+theorem one_lt_of_one_lt_of_mk_lt (ha : 1 < a) (hab : mk a < mk (b / a)) :
+    1 < b := by
+  suffices a⁻¹ < b / a by
+    simpa using this
+  apply lt_of_mk_lt_mk_of_le_one
+  · simpa using hab
+  · simpa using ha.le
+
+@[to_additive archimedean_of_mk_eq_mk]
+theorem mulArchimedean_of_mk_eq_mk (h : ∀ (a b : M), a ≠ 1 → b ≠ 1 → mk a = mk b) :
+    MulArchimedean M where
+  arch := by
+    intro x y hy
+    by_cases hx : x ≤ 1
+    · use 0
+      simpa using hx
+    · have hx : 1 < x := lt_of_not_ge hx
+      have hxy : mk x = mk y := h x y hx.ne.symm hy.ne.symm
+      obtain ⟨⟨m, hm⟩, _⟩ := (MulArchimedeanClass.eq).mp hxy
+      rw [mabs_eq_self.mpr hx.le, mabs_eq_self.mpr hy.le] at hm
+      exact ⟨m, hm⟩
+
+@[to_additive mk_eq_mk_of_archimedean]
+theorem mk_eq_mk_of_mulArchimedean [MulArchimedean M] (ha : a ≠ 1) (hb : b ≠ 1) :
+    mk a = mk b := by
+  obtain hm := MulArchimedean.arch |a|ₘ (show 1 < |b|ₘ by simpa using hb)
+  obtain hn := MulArchimedean.arch |b|ₘ (show 1 < |a|ₘ by simpa using ha)
+  exact eq.mpr ⟨hm, hn⟩
+
+section Hom
+
+variable {N : Type*} [CommGroup N] [LinearOrder N] [IsOrderedMonoid N]
+variable {F : Type*} [FunLike F M N] [OrderHomClass F M N] [MonoidHomClass F M N]
+
+/-- An ordered group homomorphism can be lifted to an OrderHom
+over archimedean classes. -/
+@[to_additive "An ordered group homomorphism can be lifted to an OrderHom
+over archimedean classes."]
+noncomputable
+def orderHom (f : F) : MulArchimedeanClass M →o MulArchimedeanClass N where
+  toFun := Quotient.map f (fun a b h ↦ by
+    obtain ⟨⟨m, hm⟩, ⟨n, hn⟩⟩ := h
+    refine ⟨⟨m, ?_⟩, ⟨n, ?_⟩⟩ <;> rw [← map_mabs, ← map_mabs, ← map_pow]
+    · exact OrderHomClass.monotone f hm
+    · exact OrderHomClass.monotone f hn
+  )
+  monotone' a b h := by
+    rw [← out_eq a, ← out_eq b] at h ⊢
```

Comment:
I'd argue this would be better as
```suggestion
    induction a using ind with | mk a
    induction b using ind with | mk b
```
after adding
```lean-4
/-- An induction principle for `MulArchimedeanClass`. -/
@[elab_as_elim]
theorem ind {motive : MulArchimedeanClass M → Prop} (mk : ∀ a, motive (.mk a)) : ∀ x, motive x := Quotient.ind mk
```
I'd encourage you to use this style as much as possible, it may simplify things.